DECISION-RECORD.md

```diff
@@ -0,0 +1,292 @@
+
+#### Cons
+
+- Harder for beginners (Lucene syntax)
+- Full-text search doesn’t make sense for all resources. E.g. Fuzziness and term boosting doesn’t make sense for time series metric data.
+- Requires URL percent encoding i.e. use of cURL or Postman becomes more complicated.
+
+---
+
+## Badges API
+
+**Decision: Implemented as a top level API endpoint during prototyping, removed from scope for feedback round**
+
+History links:
+
+- https://phabricator.wikimedia.org/T65486
+- https://gerrit.wikimedia.org/r/c/mediawiki/extensions/Wikibase/+/125886
+- https://global-search.toolforge.org/?q=availablebadges&namespaces=2%2C4%2C8&title=%28Gadgets-definition%7C.*%5C.%28js%7Ccss%7Cjson%29%29 (not used in mid 2020)
+- https://codesearch.wmcloud.org/search/?q=wbavailablebadges&i=fosho&files=&repos= (not used in mid 2020)
+
+---
+
+## Updating resources
+
+**Decision: Use PATCH and json-patch **
+
+### Patch
+
+When deciding on the update strategy of resources in our API we took into consideration two different proposals for sending updates: Payload comparison
+
+Payload comparison: https://www.notion.so/Payload-Examples-1cbe38a24f8044fba18d107d390b97fb
+
+#### RFC 6902 (json-patch+json)
+
+- More php tooling and support (not extremely popular or regularly updated but it is an old RFC):
+  - github.com/mikemccabe/json-patch-php
+  - github.com/swaggest/json-diff 
+- Explicit approach with predefined available actions
+- Allows appending to arrays
+- Multiple operations are performed by sending an array of “operation” objects
+
+#### RFC 7396 (merge-patch+json)
+
+- Straightforward, minimal verbosity
+- Removal is “implicit”
+  - i.e. by sending a key with a `null` value
+- No support for extending array values 
+  - i.e. arrays must be fully replaced
+- PHP tooling seems to be scarce and mostly abandoned
+- Multiple operations are performed by sending one “update” object
+
+---
+
+## Entities + Statements
+
+**Decision: Statements would be exposed both within an entity and also at the top level**
+
+Considered options:
+
+- For the statements with a GUID:
+  - Should we address those always through entities?
+    - longer path
+    - the entity id is already in the path and the prefix is redundant
+- Or individually, through a /statements path? 
+
+Usefulness (end users) Vs. cost of ownership (devs)
+
+### Notes
+
+The current Action API allows you to get a statement with only the GUID (without the entity id prefix) with no extra work
+
+Not having a top level statement endpoint would mean that “extra work” would be needed by the client in order to access a statement (splitting the statement ID)
+
+Taking this into account we can cover all of our existing APIs with REST APIs only if we do have statements as a top level api module.
+
+---
+
+## RevisionIds (Path? Query?)
+
+**Decision: We will not worry about revisionids in paths, as the current API doesn't expose them and per core behaviour this should probably be left for the revisions endpoint**
```

Comment:
Is this what we decided here? I thought eventually the revision id is included with conditional requests in the ETag header for GET requests and as baserevid in the query (or maybe body?) for editing endpoints, no?

Reply:
the use of revids in conditional requests as part of the ETag does not cover the case of retrieving old revision ids.
I remember this coming up in conversation, but I dont see it in any of the docs now looking back.